README.md
# vip-8

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2021 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-01-04
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-01-04
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/vip-8/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/vip-8/compare/0.1.0...0.1.1


test/vip_8/cpu_test.clj
(ns vip-8.cpu-test
  (:require [clojure.test :refer :all]
            [vip-8.cpu :refer :all]
            [vip-8.screen :as screen]))

(deftest read-instruction-test
  (testing "instructions can be decoded"
    (is (= (read-instruction {:memory [0x00 0xe0]
                              :registers {:pc 0x000}})
           {:instruction 0x00e0
            :x 0x0
            :y 0x0
            :n 0x0
            :nn 0x0
            :nnn 0x0}))
    (is (= (read-instruction {:memory [0x12 0x24]
                              :registers {:pc 0x000}})
           {:instruction 0x1
            :x 0x0
            :y 0x0
            :n 0x0
            :nn 0x0
            :nnn 0x224}))
    (is (= (read-instruction {:memory [0x1F 0x8A]
                              :registers {:pc 0x000}})
           {:instruction 0x1
            :x 0x0
            :y 0x0
            :n 0x0
            :nn 0x0
            :nnn 0xF8A}))
    (is (= (read-instruction {:memory [0x1F 0x8A 0x00 0xEE]
                              :registers {:pc 0x002}})
           {:instruction 0x00EE
            :x 0x0
            :y 0x0
            :n 0x0
            :nn 0x0
            :nnn 0x000}))
    (is (= (read-instruction {:memory [0x1F 0x8A 0x23 0x9A]
                              :registers {:pc 0x002}})
           {:instruction 0x2
            :x 0x0
            :y 0x0
            :n 0x0
            :nn 0x0
            :nnn 0x39A}))
    (is (= (read-instruction {:memory [0x1F 0x8A 0xA1 0x0B]
                              :registers {:pc 0x002}})
           {:instruction 0xA
            :x 0x0
            :y 0x0
            :n 0x0
            :nn 0x0
            :nnn 0x10B})))
  (is (= (read-instruction {:memory [0x1F 0x8A 0xA1 0x0B 0x60 0x45]
                            :registers {:pc 0x004}})
         {:instruction 0x6
            :x 0x0
            :y 0x0
            :n 0x0
            :nn 0x45
            :nnn 0x000}))
  (is (= (read-instruction {:memory [0x63 0xCD]
                            :registers {:pc 0x000}})
         {:instruction 0x6
            :x 0x3
            :y 0x0
            :n 0x0
            :nn 0xCD
            :nnn 0x000})))

(deftest execute-test
  (testing "instruction 0x00e0 clears the screen"
    (let [ cleared? (atom false)]
      (with-redefs [screen/clear 
                    (fn [] 
                      (swap! cleared? (fn [_] true)))]
        (execute {:instruction 0x00e0} {})
        (is @cleared?))))
  (testing "instruction 0xA sets the index registry"
    (let [ cleared? (atom false)]
      (with-redefs [screen/clear 
                    (fn [] 
                      (swap! cleared? (fn [_] true)))]
        (let [result 
              (execute {:instruction 0xA
                        :nnn 0x001} 
                       {:memory [0x1A 0xBA 0x91] 
                        :registers {:index 0x0}})]
          (is (not  @cleared?))
          (is (= (:index (:registers result))
                 0xBA91)))
        (let [result 
              (execute {:instruction 0xA
                        :nnn 0x000} 
                       {:memory [0x1C 0xBA] 
                        :registers {:index 0x0}})]
          (is (not  @cleared?))
          (is (= (:index (:registers result))
                 0x1CBA))))))
  (testing "instruction 0x6 sets given register to given value"
    (let [status {:registers {:v0 0x0
                              :v1 0x0
                              :v2 0x0
                              :v3 0x0
                              :v4 0x0
                              :v5 0x0
                              :v6 0x0
                              :v7 0x0
                              :v8 0x0
                              :v9 0x0
                              :vA 0x0
                              :vB 0x0
                              :vC 0x0
                              :vD 0x45
                              :vE 0x69
                              :vF 0x0}}]
      (is (= (:v3 (:registers (execute {:instruction 0x6
                                        :x 0x3
                                        :y 0x0
                                        :n 0x0
                                        :nn 0x15
                                        :nnn 0x0}
                                       status)))
             0x15))
      (is (= (:v3 (:registers (execute {:instruction 0x6
                                        :x 0x3
                                        :y 0x0
                                        :n 0x0
                                        :nn 0xBA
                                        :nnn 0x0}
                                       status)))
             0xBA))
      (let [actual (execute {:instruction 0x6
                                        :x 0xD
                                        :y 0x0
                                        :n 0x0
                                        :nn 0xFF
                                        :nnn 0x0}
                                       status)
            registers (:registers actual)]
        (is (= (:v3 registers) 0x0))
        (is (= (:vD registers) 0xFF))))))


test/vip_8/rom_test.clj
(ns vip-8.rom-test
  (:require [clojure.test :refer :all]
            [clojure.java.io :refer [file input-stream output-stream]]
            [vip-8.rom :refer :all]))

(deftest load-rom-test
  (testing "A ROM cam be loaded"
    (let [file (java.io.File/createTempFile "test" nil)
          expected [0 224 42 64]]
      (with-open [out (output-stream file)]
        (loop [rbytes expected]
          (if (not-empty rbytes)
            (let [next-byte (first rbytes)]
              (.write out next-byte)
              (recur (rest rbytes)))
            nil)))
      (is (= expected
             (read-rom (.getAbsolutePath file)))))))


test/vip_8/core_test.clj
(ns vip-8.core-test
  (:require [clojure.test :refer [deftest testing is]]
            [vip-8.core :refer :all]
            [vip-8.rom :as rom]
            [vip-8.screen :as screen]))

(deftest load-rom-tests
  (testing "font is loading in memory in addresses from 050 to 09F"
    (with-redefs [rom/read-rom (fn [_] [])]
      (let [status (load-rom "test.ch8")
            memory (:memory status)]
        (is (= (nth memory 0x050)
               0xF0))
        (is (= (nth memory 0x051)
               0x90))
        (is (= (nth memory 0x052)
               0x90))
        (is (= (nth memory 0x053)
               0x90))
        (is (= (nth memory 0x054)
               0xF0))
        (is (= (nth memory 0x055)
               0x20))
        (is (= (nth memory 0x056)
               0x60))
        (is (= (nth memory 0x057)
               0x20))
        (is (= (nth memory 0x058)
               0x20))
        (is (= (nth memory 0x59)
               0x70))
        (is (= (nth memory 0x5A)
               0xF0))
        (is (= (nth memory 0x5B)
               0x10))
        (is (= (nth memory 0x5C)
               0xF0))
        (is (= (nth memory 0x5D)
               0x80))
        (is (= (nth memory 0x5E)
               0xF0))
        (is (=  (nth memory 0x5F)
               0xF0))
        (is (= (nth memory 0x60)
               0x10))
        (is (= (nth memory 0x61)
               0xF0))
        (is (= (nth memory 0x62)
               0x10))
        (is (= (nth memory 0x63)
               0xF0))
        (is (= (nth memory 0x64)
               0x90))
        (is (= (nth memory 0x65)
               0x90))
        (is (= (nth memory 0x66)
               0xF0))
        (is (= (nth memory 0x67)
               0x10))
        (is (= (nth memory 0x68)
               0x10))
        (is (= (nth memory 0x69)
               0xF0))
        (is (= (nth memory 0x6A)
               0x80))
        (is (= (nth memory 0x6B)
               0xF0))
        (is (= (nth memory 0x6C)
               0x10))
        (is (= (nth memory 0x6D)
               0xF0))
        (is (= (nth memory 0x6E)
               0xF0))
        (is (= (nth memory 0x6F)
               0x80))
        (is (= (nth memory 0x70)
               0xF0))
        (is (= (nth memory 0x71)
               0x90))
        (is (= (nth memory 0x72)
               0xF0))
        (is (= (nth memory 0x73)
               0xF0))
        (is (= (nth memory 0x74)
               0x10))
        (is (= (nth memory 0x75)
               0x20))
        (is (= (nth memory 0x76)
               0x40))
        (is (= (nth memory 0x77)
               0x40))
        (is (= (nth memory 0x78)
               0xF0))
        (is (= (nth memory 0x79)
               0x90))
        (is (= (nth memory 0x7A)
               0xF0))
        (is (= (nth memory 0x7B)
               0x90))
        (is (= (nth memory 0x7C)
               0xF0))
        (is (= (nth memory 0x7D)
               0xF0))
        (is (= (nth memory 0x7E)
               0x90))
        (is (= (nth memory 0x7F)
               0xF0))
        (is (= (nth memory 0x80)
               0x10))
        (is (= (nth memory 0x81)
               0xF0))
(is (= (nth memory 0x82)
       0xF0))
(is (= (nth memory 0x83)
       0x90))
(is (= (nth memory 0x84)
       0xF0))
(is (= (nth memory 0x85)
       0x90))
(is (= (nth memory 0x86)
       0x90))
(is (= (nth memory 0x87)
       0xE0))
(is (= (nth memory 0x88)
       0x90))
(is (= (nth memory 0x89)
       0xE0))
(is (= (nth memory 0x8A)
       0x90))
(is (= (nth memory 0x8B)
       0xE0))
(is (= (nth memory 0x8C)
       0xF0))
(is (= (nth memory 0x8D)
       0x80))
(is (= (nth memory 0x8E)
       0x80))
(is (= (nth memory 0x8F)
       0x80))
(is (= (nth memory 0x90)
       0xF0))
(is (= (nth memory 0x91)
       0xE0))
(is (= (nth memory 0x92)
       0x90))
(is (= (nth memory 0x93)
       0x90))
(is (= (nth memory 0x94)
       0x90))
(is (= (nth memory 0x95)
       0xE0))
(is (= (nth memory 0x96)
       0xF0))
(is (= (nth memory 0x97)
       0x80))
(is (= (nth memory 0x98)
       0xF0))
(is (= (nth memory 0x99)
       0x80))
(is (= (nth memory 0x9A)
       0xF0))
(is (= (nth memory 0x9B)
       0xF0))
(is (= (nth memory 0x9C)
       0x80))
(is (= (nth memory 0x9D)
       0xF0))
(is (= (nth memory 0x9E)
       0x80))
(is (= (nth memory 0x9F)
       0x80)))))
(testing "Rom is loaded from address 0x200"
  (let [expected [0x00 0xE0 0x14 0x78]]
    (with-redefs [rom/read-rom (fn [_] expected)]
      (let [status (load-rom "test.ch8")
            memory (:memory status)
            actual (take (count expected) 
                         (drop 0x200 memory))]
        (is (= actual expected))))))
(testing "there are 4096 bytes of available memory"
  (let [rom [0x00 0xE0 0x14 0x78]]
    (with-redefs [rom/read-rom (fn [_] rom)]
      (let [status (load-rom "test.ch8")
            memory (:memory status)]
        (is (= 4096 (count memory)))))))
(testing "the stack is empty"
  (with-redefs [rom/read-rom (fn [_] [])]
    (let [stack (:stack (load-rom "test.ch8"))]
      (is (and (empty? stack)
               (not (nil? stack)))))))
(testing "the v registers are set to 0x00"
  (with-redefs [rom/read-rom (fn [_] [])]
    (let [registers (:registers (load-rom "tetris.ch8"))]
      (is (and (= (:v0 registers)
                  0x00)
               (= (:v1 registers)
                  0x00)
               (= (:v2 registers)
                  0x00)
               (= (:v3 registers)
                  0x00)
               (= (:v4 registers)
                  0x00)
               (= (:v5 registers)
                  0x00)
               (= (:v6 registers)
                  0x00)
               (= (:v7 registers)
                  0x00)
               (= (:v8 registers)
                  0x00)
               (= (:v9 registers)
                  0x00)
               (= (:vA registers)
                  0x00)
               (= (:vB registers)
                  0x00)
               (= (:vC registers)
                  0x00)
               (= (:vD registers)
                  0x00)
               (= (:vE registers)
                  0x00)
               (= (:vF registers)
                  0x00))))))
(testing "program counter is set to 0x200"
  (with-redefs [rom/read-rom (fn [_] [])]
    (is (= (:pc (:registers (load-rom "tetris.ch8")))
           0x200))))
(testing "timers start at value 0"
  (with-redefs [rom/read-rom (fn [_] [])]
    (let [timers (:timers (load-rom "tetris.ch8"))]
      (is (and (= (:delay timers)
                  0)
               (= (:sound timers)
                  0))))))
(testing "index register is set to 0"
  (with-redefs [rom/read-rom (fn [_] [])]
    (let [registers (:registers (load-rom "tetris.ch8"))]
      (is (= (:index registers) 0x000)))))
(let [screen (atom (repeat 32 (repeat 64 true)))]
  (with-redefs [rom/read-rom 
                (fn [_] [0x00 0xE0 0xA2 0x01])
                screen/clear 
                (fn []
                  (swap! screen 
                         (fn [_] 
                           (repeat 32 (repeat 64 false)))))]
    (let [initial-status (load-rom "pong.ch8")]
      (testing "clear instruction clears the screen"
        (let [status (step initial-status)]
          (is (= (:pc (:registers status))
                 0x202))
          (is (= (count (filter 
                          #(> (count 
                                (filter (fn [x] x) %)) 0) @screen))
                 0))))
      (testing "index register can be set"
        (let [status (step (assoc initial-status
                                  :registers
                                  (assoc (:registers initial-status)
                                         :pc
                                         0x202)))
              registers (:registers status)]
          (is (= (:pc registers) 0x204))
          (is (= (:index registers) 0xE0A2))))))))


src/vip_8/cpu.clj
(ns vip-8.cpu
  (:require [vip-8.screen :as screen]
            [clojure.string :as string]))

(defn read-instruction 
  "Reads the instruction at the program counter
  and returns the nibles"
  [status]
  (let [pc (:pc (:registers status))
        memory (:memory status)
        first-byte (nth memory pc)
        second-byte (nth memory (inc pc))
        i (bit-or (bit-shift-left first-byte 0x8)
                  second-byte)
        opcode (bit-shift-right (bit-and i 0xF000) 0xC)]
    (cond
      (= opcode 0x0)
      {:instruction i
       :x 0
       :y 0
       :n 0
       :nn 0
       :nnn 0}
      (= opcode 0x6)
      {:instruction opcode
       :x (bit-and first-byte 0xF)
       :y 0
       :n 0
       :nn second-byte
       :nnn 0x0}
      :else
      {:instruction opcode
       :x 0
       :y 0
       :n 0
       :nn 0
       :nnn (bit-and i 0xFFF)})))

(defn execute 
  "Executes a given instruction"
  [instruction state]
  (defn clear-screen []
    (screen/clear)
    state)

  (defn set-register [status register value]
    (assoc status
           :registers
           (assoc (:registers status)
                  register
                  value)))

  (defn set-v-register []
    (set-register state
                  (keyword (str "v" (string/upper-case (format "%x" (:x instruction)))))
                  (:nn instruction)))

  (defn set-index-register []
    (let [address (:nnn instruction)
          memory (:memory state)
          first-byte (nth memory address)
          second-byte (nth memory (inc address))
          new-value (bit-or (bit-shift-left first-byte 0x8)
                            second-byte)]
      (set-register state
                    :index
                    new-value)))

  (let [instructions {:e0 clear-screen
                      :6 set-v-register
                      :a set-index-register}
        op-keyword (keyword (format "%x" (:instruction instruction)))]
  (if (nil? op-keyword)
    (throw "Instruction ")
    ((op-keyword instructions)))))



src/vip_8/core.clj
(ns vip-8.core
  (:require [vip-8.rom :as rom]
            [vip-8.screen :as screen]
            [vip-8.cpu :as cpu]))

(defn load-rom [filename]
  (let [font '(
               0xF0 0x90 0x90 0x90 0xF0 
               0x20 0x60 0x20 0x20 0x70
               0xF0 0x10 0xF0 0x80 0xF0
               0xF0 0x10 0xF0 0x10 0xF0
               0x90 0x90 0xF0 0x10 0x10
               0xF0 0x80 0xF0 0x10 0xF0
               0xF0 0x80 0xF0 0x90 0xF0
               0xF0 0x10 0x20 0x40 0x40
               0xF0 0x90 0xF0 0x90 0xF0
               0xF0 0x90 0xF0 0x10 0xF0
               0xF0 0x90 0xF0 0x90 0x90
               0xE0 0x90 0xE0 0x90 0xE0
               0xF0 0x80 0x80 0x80 0xF0
               0xE0 0x90 0x90 0x90 0xE0
               0xF0 0x80 0xF0 0x80 0xF0
               0xF0 0x80 0xF0 0x80 0x80)
        rom (rom/read-rom filename)]
    {:memory (into [] (concat (repeat 0x050 0x00) 
                              font
                              (repeat (- 0x200 (+ 0x50 (count font))) 
                                      0x00)
                              rom
                              (repeat (- 4096 (+ 0x200 (count rom)))
                                      0x00)))
     :stack []
     :registers {:v0 0x00
                 :v1 0x00
                 :v2 0x00
                 :v3 0x00
                 :v4 0x00
                 :v5 0x00
                 :v6 0x00
                 :v7 0x00
                 :v8 0x00
                 :v9 0x00
                 :vA 0x00
                 :vB 0x00
                 :vC 0x00
                 :vD 0x00
                 :vE 0x00
                 :vF 0x00
                 :pc 0x200
                 :index 0x000}
     :timers {:delay 0
              :sound 0}}))

(defn step
  "Executes an instruction and returns the resulting status"
  [prev-status]
  (let [instruction 
        (cpu/read-instruction prev-status)
        after-reading-status 
    (assoc prev-status 
           :registers 
           (assoc (:registers prev-status)
                  :pc
                  (+ (:pc (:registers prev-status)) 2)))]
    (cpu/execute instruction 
                 after-reading-status)))
